src/main/java/com/thomashan/collection/immutable/ImmutableListImpl.java

```diff
@@ -0,0 +1,188 @@
+package com.thomashan.collection.immutable;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+
+/* default */
+final class ImmutableListImpl<E> implements ImmutableList<E> {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [This class has too many methods, consider refactoring it.](https://app.codacy.com/app/thomashan/coup-java/pullRequest?prid=4509131)